third_party/move/move-model/src/builder/module_builder.rs

```diff
@@ -875,6 +875,95 @@ impl<'env, 'translator> ModuleBuilder<'env, 'translator> {
 /// # Definition Analysis
 
 impl<'env, 'translator> ModuleBuilder<'env, 'translator> {
+    pub fn check_language_version(&self, loc: &Loc, feature: &str, version_min: LanguageVersion) {
+        if !self.parent.env.language_version().is_at_least(version_min) {
+            self.parent.env.error(
+                loc,
+                &format!(
+                    "not supported before language version `{}`: {}",
+                    version_min, feature
+                ),
+            )
+        }
+    }
+
+    fn analyze_constant(
+        &mut self,
+        key: &PA::ConstantName,
+        constant_map: &UniqueMap<PA::ConstantName, EA::Constant>,
+        visiting: &mut HashSet<PA::ConstantName>,
+        visited: &mut HashSet<PA::ConstantName>,
+        compiled_module: &Option<BytecodeModule>,
+    ) {
+        let qsym = self.qualified_by_module_from_name(&key.0);
+        let loc = self
+            .parent
+            .const_table
+            .get(&qsym)
+            .expect("constant declared")
+            .loc
+            .clone();
+        if visited.contains(key) {
+            return;
+        }
+        if visiting.contains(key) {
+            return;
+        }
+        visiting.insert(*key);
+        if let Some(exp) = constant_map.get(key) {
+            let names = exp.value.value.get_names_for_const_exp();
+            for name in names {
+                let const_name = PA::ConstantName(name);
+                let qsym = self.qualified_by_module_from_name(&name);
+                if !self.parent.const_table.contains_key(&qsym) {
+                    continue;
```

Comment:
Why is this a continue and not an error?

Reply:
In `def_ana_constant`, `is_valid_for_constant` is called which can give a better error message.

Reply:
Is it really a better error message?  I think it would be better if we could say something like "Z is not a constant", rather than 
```
error: Invalid expression in `const`. Constant folding failed due to incomplete evaluation
  ┌─ tests/checking/typing/recursive_constant.move:8:8
  │
8 │        Z + A
  │        ^^^^^
```
I regret that I never realized that the constant names were getting inlined into the code as `Value`, rather than having a more explicit lookup of a constant.  I was too new when I did the constant folding.

Or, as V1 does:
```
error[E04013]: invalid statement or expression in constant
   ┌─ tests/move_check/typing/constant_unsupported_exps.move:45:9
   │
45 │         FLAG;
   │         ^^^^ Other constants are not supported in constants
```

Reply:
In this test case, both `Z` and `A` are constants but we cannot compute the value because of cyclic dependency. So I think this error message makes sense to me.